### DISCUSSION TITLE
Need ability to disable stack in _data error response to limit attackers identifying and exploiting dependencies

### DISCUSSION THREAD

**Discussion by ngbrown** [remix-run/remix#5199] [Proposals]
### What version of Remix are you using?

1.5.1

### Steps to Reproduce

Pass an invalid parameter to the `/?_data=routes%2Fsomething_invalid` endpoint. Observe the results. Or pass an invalid POST body to delve deeper into the application stack.

### Expected Behavior

Need a way to disable the stack return data in production and have a more generic error returned to the client.

### Actual Behavior

During testing by our security analyst, a stack trace was returned that identified the logging system in use and the open-source libraries of the application code running.  This could be used by an attacker to exploit dependencies in the application the most serious of which may have zero day implications.

This line unconditionally returns the stack:

https://github.com/remix-run/remix/blob/252ea6c5c55650675b771bf8dada10f68a3cabb4/packages/remix-server-runtime/errors.ts#L68

**Comment by pkalisiewicz** [comment]
I'm happy to work on this bug, the question is, do we want to just remove the stack trace when in production mode, or do we want to filter out paths, etc?

**Comment by p-m-p** [comment]
This appears to be quite a big security issue if there is no way to suppress the stack trace in responses with uncaught errors relying on use of `ErrorBoundary`. 

I need to solve this myself either through a change to the framework or by other means; perhaps adding a middleware in express to redact the stack traces.

I had a quick look and I think [`serializeError`](https://github.com/remix-run/remix/blob/5c868dceaa542d0ac61e742f62f7fc08d9f18de2/packages/remix-server-runtime/errors.ts#L67) could possibly take a second parameter to optionally add the stack trace if `serverMode` is not set to `ServerMode.Production` in the [handler functions](https://github.com/remix-run/remix/blob/5c868dceaa542d0ac61e742f62f7fc08d9f18de2/packages/remix-server-runtime/server.ts#L691)?

**Comment by kasprownik** [comment]
Hey! If you want to quick fix the above error, this is a snippet from the express instance I use to strip stack from the response and return 400 on incorrect `_data` param:

```js
app.all(
  '*',
  createRequestHandler({ build: require('../build'), mode: 'production' }),
  (err, req, res, next) => {
    if (
      err.message === "Cannot read properties of undefined (reading 'params')"
    ) {
      res.status(400).send({ error: 'bad request' });
    } else {
      res.status(500).send({ error: 'unexpected error occured' });
    }
  }
);
```

**Comment by p-m-p** [comment]
> Hey! If you want to quick fix the above error, this is a snippet from the express instance I use to strip stack from the response and return 400 on incorrect `_data` param:
> 
> ```js
> app.all(
>   '*',
>   createRequestHandler({ build: require('../build'), mode: 'production' }),
>   (err, req, res, next) => {
>     if (
>       err.message === "Cannot read properties of undefined (reading 'params')"
>     ) {
>       res.status(400).send({ error: 'bad request' });
>     } else {
>       res.status(500).send({ error: 'unexpected error occured' });
>     }
>   }
> );
> ```

Probably not going to work so well if you want to maintain use of `CatchBoundary` for things like 404 page. 

It's probably *possible* to remove stack from data and page requests (it's in appState) with a middleware and a bit of hackery. I've not looked into it yet but imagine there will be some trade offs in being able to use server components/streams.

**Comment by machour** [comment]
Converting this to a Proposal discussion, as per our [Development Process](https://remix.run/docs/en/v1/pages/contributing#development-process).

**Reply by p-m-p** [reply]
I had raised this proposal previously https://github.com/remix-run/remix/discussions/4564.

**Comment by ngbrown** [comment]
Recent versions of Remix now include a `serializeErrors()` function that also includes this problem.

I noticed that `react-router-dom` server.js's `serializeErrors()` now includes a [line](https://github.com/remix-run/react-router/blob/f1157a0553498957e06cdd081f38240c778d70bf/packages/react-router-dom/server.tsx#L157):

> // Do not serialize stack traces from SSR for security reasons

Other than that, it is the exact same function. So the solution looks to be to copy from `react-router-dom` to `@remix-run/server-runtime` and also remove `stack` from `serializeError`.

**Comment by brophdawg11** [comment]
This should be resolved as of `1.15.0` via https://github.com/remix-run/remix/pull/5541.  There is also a new [`handleError`](https://remix.run/docs/en/main/file-conventions/entry.server#handleerror) function where you can capture and log/report the error on the server prior to the stack trace being stripped.
